core/drivers/crypto/versal/pki/ecc_pki_engine.c

```diff
@@ -0,0 +1,853 @@
+		   PKI_RQ_CFG_PAGE_SIZE_1024);
+	io_write32(versal_pki.regs + PKI_RQ_CFG_CQID_OFFSET, PKI_RQ_CFG_CQID);
+	io_write32(versal_pki.regs + PKI_CQ_CFG_SIZE_OFFSET,
+		   PKI_CQ_CFG_SIZE_4096);
+	io_write32(versal_pki.regs + PKI_CQ_CFG_IRQ_IDX_OFFSET,
+		   PKI_CQ_CFG_IRQ_ID_VAL);
+	io_write32(versal_pki.regs + PKI_RQ_CFG_QUEUE_DEPTH_OFFSET,
+		   PKI_RQ_CFG_QUEUE_DEPTH_VAL);
+	io_write64(versal_pki.regs + PKI_CRYPTO_IRQ_ENABLE_OFFSET,
+		   PKI_IRQ_ENABLE_VAL);
+
+	io_write32(versal_pki.regs + PKI_CQ_CTL_TRIGPOS_OFFSET,
+		   PKI_CQ_CTL_TRIGPOS_VAL);
+	io_write64(versal_pki.regs + PKI_RQ_CTL_NEW_REQUEST_OFFSET, reqval);
+
+	/* Wait for completion */
+	while (retries--) {
+		uint64_t irq_status = io_read64(versal_pki.regs +
+						PKI_CRYPTO_IRQ_STATUS_OFFSET);
+		if (irq_status == PKI_IRQ_DONE_STATUS_VAL) {
+			io_write64(versal_pki.regs +
+				   PKI_CRYPTO_IRQ_RESET_OFFSET,
+				   PKI_IRQ_DONE_STATUS_VAL);
+			ret = TEE_SUCCESS;
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static TEE_Result pki_check_status(void)
+{
+	uint32_t cq_status = 0;
+	uint32_t cq_value = 0;
+
+	cache_operation(TEE_CACHEFLUSH, versal_pki.cq, PKI_QUEUE_BUF_SIZE);
+
+	cq_status = io_read32((vaddr_t)versal_pki.cq);
+	cq_value = io_read32((vaddr_t)versal_pki.cq + 4);
+
+	if (cq_status != PKI_EXPECTED_CQ_STATUS ||
+	    cq_value != PKI_EXPECTED_CQ_VALUE)
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_ecc_verify(uint32_t algo, struct ecc_public_key *key,
+			     const uint8_t *msg, size_t msg_len,
+			     const uint8_t *sig, size_t sig_len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	size_t bits = 0;
+	size_t bytes = 0;
+	size_t len = 0;
+	uintptr_t addr = (uintptr_t)versal_pki.rq_in;
+
+	ret = versal_ecc_get_key_size(key->curve, &bytes, &bits);
+	if (ret)
+		return ret;
+
+	/* Copy public key */
+	versal_crypto_bignum_bn2bin_eswap(key->curve, key->x, (uint8_t *)addr);
+	addr += bytes;
+	versal_crypto_bignum_bn2bin_eswap(key->curve, key->y, (uint8_t *)addr);
+	addr += bytes;
+
+	/* Copy signature */
+	memcpy_swp((uint8_t *)addr, sig, sig_len / 2);
+	addr += sig_len / 2;
+	memcpy_swp((uint8_t *)addr, sig + sig_len / 2, sig_len / 2);
+	addr += sig_len / 2;
+
+	/* Copy hash */
+	ret = versal_ecc_prepare_msg(algo, msg, msg_len, &len, (uint8_t *)addr);
+	if (ret)
+		return ret;
+	if (len < bytes)
+		memset((uint8_t *)addr + len, 0, bytes - len);
+	addr += bytes;
+
+	if (key->curve == TEE_ECC_CURVE_NIST_P521) {
+		memset((uint8_t *)addr, 0, PKI_VERIFY_P521_PADD_BYTES);
+		addr += PKI_VERIFY_P521_PADD_BYTES;
+	}
+
+	/* Build descriptors */
+	ret = pki_build_descriptors(key->curve, PKI_DESC_OPTYPE_ECDSA_VERIFY,
+				    (uint32_t *)addr);
+	if (ret)
+		return ret;
+
+	ret = pki_start_operation(PKI_NEW_REQUEST_MASK & (addr + 1));
+	if (ret)
+		return ret;
+
+	ret = pki_check_status();
+	if (ret)
+		return ret;
+
+	/* Clear memory */
+	memset(versal_pki.rq_in, 0, PKI_QUEUE_BUF_SIZE);
+	memset(versal_pki.cq, 0, PKI_QUEUE_BUF_SIZE);
```

Comment:
This is a leftover from Versal Net baremetal code. It doesn't hurt not to clear the buffers at all, except when generating a keypair to avoid leaking secrets.